File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/JdbcBufferedConsumerFactory.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.jdbc;

import static io.airbyte.cdk.integrations.base.JavaBaseConstants.DEFAULT_AIRBYTE_INTERNAL_NAMESPACE;
import static io.airbyte.cdk.integrations.destination.jdbc.AbstractJdbcDestination.RAW_SCHEMA_OVERRIDE;
import static io.airbyte.cdk.integrations.destination.jdbc.constants.GlobalDataSizeConstants.DEFAULT_MAX_BATCH_SIZE_BYTES;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Preconditions;
import io.airbyte.cdk.db.jdbc.JdbcDatabase;
import io.airbyte.cdk.db.jdbc.JdbcUtils;
import io.airbyte.cdk.integrations.base.AirbyteMessageConsumer;
import io.airbyte.cdk.integrations.base.DestinationConfig;
import io.airbyte.cdk.integrations.base.TypingAndDedupingFlag;
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer;
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.BufferedStreamConsumer;
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.OnCloseFunction;
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.OnStartFunction;
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.RecordWriter;
import io.airbyte.cdk.integrations.destination.record_buffer.InMemoryRecordBufferingStrategy;
import io.airbyte.commons.json.Jsons;
import io.airbyte.integrations.base.destination.typing_deduping.StreamId;
import io.airbyte.integrations.base.destination.typing_deduping.TyperDeduper;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import io.airbyte.protocol.models.v0.AirbyteRecordMessage;
import io.airbyte.protocol.models.v0.AirbyteStream;
import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
import io.airbyte.protocol.models.v0.DestinationSyncMode;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Strategy:
 * <p>
 * 1. Create a final table for each stream
 * <p>
 * 2. Accumulate records in a buffer. One buffer per stream
 * <p>
 * 3. As records accumulate write them in batch to the database. We set a minimum numbers of records
 * before writing to avoid wasteful record-wise writes. In the case with slow syncs this will be
 * superseded with a periodic record flush from {@link BufferedStreamConsumer#periodicBufferFlush()}
 * <p>
 * 4. Once all records have been written to buffer, flush the buffer and write any remaining records
 * to the database (regardless of how few are left)
 */
public class JdbcBufferedConsumerFactory {

  private static final Logger LOGGER = LoggerFactory.getLogger(JdbcBufferedConsumerFactory.class);

  public static AirbyteMessageConsumer create(final Consumer<AirbyteMessage> outputRecordCollector,
                                              final JdbcDatabase database,
                                              final SqlOperations sqlOperations,
                                              final NamingConventionTransformer namingResolver,
                                              final JsonNode config,
                                              final ConfiguredAirbyteCatalog catalog) {
    return create(outputRecordCollector, database, sqlOperations, namingResolver, config, catalog, null);
  }

  public static AirbyteMessageConsumer create(final Consumer<AirbyteMessage> outputRecordCollector,
                                              final JdbcDatabase database,
                                              final SqlOperations sqlOperations,
                                              final NamingConventionTransformer namingResolver,
                                              final JsonNode config,
                                              final ConfiguredAirbyteCatalog catalog,
                                              final TyperDeduper typerDeduper) {
    final List<WriteConfig> writeConfigs = createWriteConfigs(namingResolver, config, catalog, sqlOperations.isSchemaRequired());

    return new BufferedStreamConsumer(
        outputRecordCollector,
        onStartFunction(database, sqlOperations, writeConfigs, typerDeduper),
        new InMemoryRecordBufferingStrategy(recordWriterFunction(database, sqlOperations, writeConfigs, catalog), DEFAULT_MAX_BATCH_SIZE_BYTES),
        onCloseFunction(typerDeduper),
        catalog,
        sqlOperations::isValidData,
        DestinationConfig.getInstance().getTextValue("schema"));
  }

  private static List<WriteConfig> createWriteConfigs(final NamingConventionTransformer namingResolver,
                                                      final JsonNode config,
                                                      final ConfiguredAirbyteCatalog catalog,
                                                      final boolean schemaRequired) {
    if (schemaRequired) {
      Preconditions.checkState(config.has("schema"), "jdbc destinations must specify a schema.");
    }
    final Instant now = Instant.now();
    return catalog.getStreams().stream().map(toWriteConfig(namingResolver, config, now, schemaRequired)).collect(Collectors.toList());
  }

  private static Function<ConfiguredAirbyteStream, WriteConfig> toWriteConfig(
                                                                              final NamingConventionTransformer namingResolver,
                                                                              final JsonNode config,
                                                                              final Instant now,
                                                                              final boolean schemaRequired) {
    return stream -> {
      Preconditions.checkNotNull(stream.getDestinationSyncMode(), "Undefined destination sync mode");
      final AirbyteStream abStream = stream.getStream();

      final String defaultSchemaName = schemaRequired ? namingResolver.getIdentifier(config.get("schema").asText())
          : namingResolver.getIdentifier(config.get(JdbcUtils.DATABASE_KEY).asText());
      // Method checks for v2
      final String outputSchema = getOutputSchema(abStream, defaultSchemaName, namingResolver);
      final String streamName = abStream.getName();
      final String tableName;
      final String tmpTableName;
      // TODO: Should this be injected from outside ?
      if (TypingAndDedupingFlag.isDestinationV2()) {
        final var finalSchema = Optional.ofNullable(abStream.getNamespace()).orElse(defaultSchemaName);
        final var rawName = StreamId.concatenateRawTableName(finalSchema, streamName);
        tableName = namingResolver.convertStreamName(rawName);
        tmpTableName = namingResolver.getTmpTableName(rawName);
      } else {
        tableName = namingResolver.getRawTableName(streamName);
        tmpTableName = namingResolver.getTmpTableName(streamName);
      }
      final DestinationSyncMode syncMode = stream.getDestinationSyncMode();

      final WriteConfig writeConfig = new WriteConfig(streamName, abStream.getNamespace(), outputSchema, tmpTableName, tableName, syncMode);
      LOGGER.info("Write config: {}", writeConfig);

      return writeConfig;
    };
  }

  /**
   * Defer to the {@link AirbyteStream}'s namespace. If this is not set, use the destination's default
   * schema. This namespace is source-provided, and can be potentially empty.
   * <p>
   * The logic here matches the logic in the catalog_process.py for Normalization. Any modifications
   * need to be reflected there and vice versa.
   */
  private static String getOutputSchema(final AirbyteStream stream,
                                        final String defaultDestSchema,
                                        final NamingConventionTransformer namingResolver) {
    if (TypingAndDedupingFlag.isDestinationV2()) {
      return namingResolver
          .getNamespace(TypingAndDedupingFlag.getRawNamespaceOverride(RAW_SCHEMA_OVERRIDE).orElse(DEFAULT_AIRBYTE_INTERNAL_NAMESPACE));
    } else {
      return namingResolver.getNamespace(Optional.ofNullable(stream.getNamespace()).orElse(defaultDestSchema));
    }
  }

  /**
   * Sets up destination storage through:
   * <p>
   * 1. Creates Schema (if not exists)
   * <p>
   * 2. Creates airybte_raw table (if not exists)
   * <p>
   * 3. <Optional>Truncates table if sync mode is in OVERWRITE
   *
   * @param database JDBC database to connect to
   * @param sqlOperations interface for execution SQL queries
   * @param writeConfigs settings for each stream
   * @param typerDeduper
   * @return
   */
  private static OnStartFunction onStartFunction(final JdbcDatabase database,
                                                 final SqlOperations sqlOperations,
                                                 final List<WriteConfig> writeConfigs,
                                                 final TyperDeduper typerDeduper) {
    return () -> {
      typerDeduper.prepareTables();
      LOGGER.info("Preparing raw tables in destination started for {} streams", writeConfigs.size());
      final List<String> queryList = new ArrayList<>();
      for (final WriteConfig writeConfig : writeConfigs) {
        // These values should be updated for destinations v2
        final String schemaName = writeConfig.getOutputSchemaName();
        final String dstTableName = writeConfig.getOutputTableName();
        LOGGER.info("Preparing raw table in destination started for stream {}. schema: {}, table name: {}",
            writeConfig.getStreamName(),
            schemaName,
            dstTableName);
        sqlOperations.createSchemaIfNotExists(database, schemaName);
        sqlOperations.createTableIfNotExists(database, schemaName, dstTableName);
        switch (writeConfig.getSyncMode()) {
          case OVERWRITE -> queryList.add(sqlOperations.truncateTableQuery(database, schemaName, dstTableName));
          case APPEND, APPEND_DEDUP -> {}
          default -> throw new IllegalStateException("Unrecognized sync mode: " + writeConfig.getSyncMode());
        }
      }
      sqlOperations.executeTransaction(database, queryList);
      LOGGER.info("Preparing raw tables in destination completed.");
    };
  }

  /**
   * Writes {@link AirbyteRecordMessage} to JDBC database's airbyte_raw table
   *
   * @param database JDBC database to connect to
   * @param sqlOperations interface of SQL queries to execute
   * @param writeConfigs settings for each stream
   * @param catalog catalog of all streams to sync
   * @return
   */
  private static RecordWriter<AirbyteRecordMessage> recordWriterFunction(final JdbcDatabase database,
                                                                         final SqlOperations sqlOperations,
                                                                         final List<WriteConfig> writeConfigs,
                                                                         final ConfiguredAirbyteCatalog catalog) {
    final Map<AirbyteStreamNameNamespacePair, WriteConfig> pairToWriteConfig = writeConfigs.stream()
        .collect(Collectors.toUnmodifiableMap(JdbcBufferedConsumerFactory::toNameNamespacePair, Function.identity()));

    return (pair, records) -> {
      if (!pairToWriteConfig.containsKey(pair)) {
        throw new IllegalArgumentException(
            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s", Jsons.serialize(catalog)));
      }

      final WriteConfig writeConfig = pairToWriteConfig.get(pair);
      sqlOperations.insertRecords(database, records, writeConfig.getOutputSchemaName(), writeConfig.getOutputTableName());
    };
  }

  /**
   * Tear down functionality
   *
   * @return
   */
  private static OnCloseFunction onCloseFunction(final TyperDeduper typerDeduper) {
    return (hasFailed) -> {
      typerDeduper.typeAndDedupe();
      typerDeduper.commitFinalTables();
      typerDeduper.cleanup();
    };
  }

  private static AirbyteStreamNameNamespacePair toNameNamespacePair(final WriteConfig config) {
    return new AirbyteStreamNameNamespacePair(config.getStreamName(), config.getNamespace());
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/JdbcBufferedConsumerFactory.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.jdbc

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.base.Preconditions
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.base.SerializedAirbyteMessageConsumer
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer
import io.airbyte.cdk.integrations.destination.StreamSyncSummary
import io.airbyte.cdk.integrations.destination.async.AsyncStreamConsumer
import io.airbyte.cdk.integrations.destination.async.buffers.BufferManager
import io.airbyte.cdk.integrations.destination.async.deser.AirbyteMessageDeserializer
import io.airbyte.cdk.integrations.destination.async.deser.IdentityDataTransformer
import io.airbyte.cdk.integrations.destination.async.deser.StreamAwareDataTransformer
import io.airbyte.cdk.integrations.destination.async.model.PartialAirbyteMessage
import io.airbyte.cdk.integrations.destination.async.state.FlushFailure
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.OnCloseFunction
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.OnStartFunction
import io.airbyte.cdk.integrations.destination.buffered_stream_consumer.RecordWriter
import io.airbyte.commons.json.Jsons
import io.airbyte.integrations.base.destination.operation.AbstractStreamOperation
import io.airbyte.integrations.base.destination.typing_deduping.ParsedCatalog
import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig
import io.airbyte.integrations.base.destination.typing_deduping.TyperDeduper
import io.airbyte.protocol.models.v0.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.*
import java.util.concurrent.Executors
import java.util.function.Consumer
import java.util.function.Function

private val LOGGER = KotlinLogging.logger {}
/**
 * Strategy:
 *
 * 1. Create a final table for each stream
 *
 * 2. Accumulate records in a buffer. One buffer per stream
 *
 * 3. As records accumulate write them in batch to the database. We set a minimum numbers of records
 * before writing to avoid wasteful record-wise writes. In the case with slow syncs this will be
 * superseded with a periodic record flush from [BufferedStreamConsumer.periodicBufferFlush]
 *
 * 4. Once all records have been written to buffer, flush the buffer and write any remaining records
 * to the database (regardless of how few are left)
 */
object JdbcBufferedConsumerFactory {

    const val DEFAULT_OPTIMAL_BATCH_SIZE_FOR_FLUSH = 25 * 1024 * 1024L

    /** @param parsedCatalog Nullable for v1 destinations. Required for v2 destinations. */
    fun createAsync(
        outputRecordCollector: Consumer<AirbyteMessage>,
        database: JdbcDatabase,
        sqlOperations: SqlOperations,
        generationIdHandler: JdbcGenerationHandler,
        namingResolver: NamingConventionTransformer,
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        defaultNamespace: String,
        typerDeduper: TyperDeduper,
        dataTransformer: StreamAwareDataTransformer = IdentityDataTransformer(),
        optimalBatchSizeBytes: Long = DEFAULT_OPTIMAL_BATCH_SIZE_FOR_FLUSH,
        parsedCatalog: ParsedCatalog,
    ): SerializedAirbyteMessageConsumer {
        if (sqlOperations.isSchemaRequired) {
            Preconditions.checkState(
                config.has("schema"),
                "jdbc destinations must specify a schema."
            )
        }
        val writeConfigs = mutableListOf<WriteConfig>()

        return AsyncStreamConsumer(
            outputRecordCollector,
            onStartFunction(
                database,
                sqlOperations,
                generationIdHandler,
                writeConfigs,
                typerDeduper,
                namingResolver,
                parsedCatalog
            ),
            onCloseFunction(
                database,
                sqlOperations,
                generationIdHandler,
                parsedCatalog,
                typerDeduper
            ),
            JdbcInsertFlushFunction(
                defaultNamespace,
                recordWriterFunction(database, sqlOperations, writeConfigs, catalog),
                optimalBatchSizeBytes
            ),
            catalog,
            BufferManager(defaultNamespace, (Runtime.getRuntime().maxMemory() * 0.2).toLong()),
            FlushFailure(),
            Executors.newFixedThreadPool(2),
            AirbyteMessageDeserializer(dataTransformer)
        )
    }

    private fun createWriteConfigs(
        database: JdbcDatabase,
        generationIdHandler: JdbcGenerationHandler,
        namingResolver: NamingConventionTransformer,
        parsedCatalog: ParsedCatalog,
    ): List<WriteConfig> {
        return parsedCatalog.streams.map {
            val rawSuffix: String =
                if (
                    it.minimumGenerationId == 0L ||
                        generationIdHandler.getGenerationIdInTable(
                            database,
                            it.id.rawNamespace,
                            it.id.rawName
                        ) == it.generationId
                ) {
                    AbstractStreamOperation.NO_SUFFIX
                } else {
                    AbstractStreamOperation.TMP_TABLE_SUFFIX
                }
            parsedStreamToWriteConfig(namingResolver, rawSuffix).apply(it)
        }
    }

    private fun parsedStreamToWriteConfig(
        namingResolver: NamingConventionTransformer,
        rawTableSuffix: String,
    ): Function<StreamConfig, WriteConfig> {
        return Function { streamConfig: StreamConfig ->
            // TODO We should probably replace WriteConfig with StreamConfig?
            // The only thing I'm not sure about is the tmpTableName thing,
            // but otherwise it's a strict improvement (avoids people accidentally
            // recomputing the table names, instead of just treating the output of
            // CatalogParser as canonical).
            WriteConfig(
                streamConfig.id.originalName,
                streamConfig.id.originalNamespace,
                streamConfig.id.rawNamespace,
                @Suppress("deprecation")
                namingResolver.getTmpTableName(streamConfig.id.rawNamespace),
                streamConfig.id.rawName,
                streamConfig.postImportAction,
                streamConfig.syncId,
                streamConfig.generationId,
                streamConfig.minimumGenerationId,
                rawTableSuffix,
            )
        }
    }

    /**
     * Sets up destination storage through:
     *
     * 1. Creates Schema (if not exists)
     *
     * 2. Creates airybte_raw table (if not exists)
     *
     * 3. <Optional>Truncates table if sync mode is in OVERWRITE
     *
     * @param database JDBC database to connect to
     * @param sqlOperations interface for execution SQL queries
     * @param writeConfigs settings for each stream </Optional>
     */
    private fun onStartFunction(
        database: JdbcDatabase,
        sqlOperations: SqlOperations,
        generationIdHandler: JdbcGenerationHandler,
        writeConfigs: MutableList<WriteConfig>,
        typerDeduper: TyperDeduper,
        namingResolver: NamingConventionTransformer,
        parsedCatalog: ParsedCatalog,
    ): OnStartFunction {
        return OnStartFunction {
            typerDeduper.prepareSchemasAndRunMigrations()
            writeConfigs.addAll(
                createWriteConfigs(database, generationIdHandler, namingResolver, parsedCatalog)
            )
            LOGGER.info {
                "Preparing raw tables in destination started for ${writeConfigs.size} streams"
            }
            val queryList: MutableList<String> = ArrayList()
            for (writeConfig in writeConfigs) {
                val schemaName = writeConfig.rawNamespace
                val dstTableName = writeConfig.rawTableName
                LOGGER.info {
                    "Preparing raw table in destination started for stream ${writeConfig.streamName}. schema: $schemaName, table name: $dstTableName"
                }
                sqlOperations.createSchemaIfNotExists(database, schemaName)
                sqlOperations.createTableIfNotExists(database, schemaName, dstTableName)
                // if rawSuffix is empty, this is a no-op
                sqlOperations.createTableIfNotExists(
                    database,
                    schemaName,
                    dstTableName + writeConfig.rawTableSuffix
                )
                when (writeConfig.minimumGenerationId) {
                    0L -> {}
                    writeConfig.generationId ->
                        if (
                            generationIdHandler.getGenerationIdInTable(
                                database,
                                schemaName,
                                dstTableName + writeConfig.rawTableSuffix
                            ) != writeConfig.generationId
                        ) {
                            queryList.add(
                                sqlOperations.truncateTableQuery(
                                    database,
                                    schemaName,
                                    dstTableName + writeConfig.rawTableSuffix,
                                )
                            )
                        }
                    else ->
                        throw IllegalStateException(
                            "Invalid minimumGenerationId ${writeConfig.minimumGenerationId} for stream ${writeConfig.streamName}. generationId=${writeConfig.generationId}"
                        )
                }
            }
            sqlOperations.executeTransaction(database, queryList)
            LOGGER.info { "Preparing raw tables in destination completed." }
            typerDeduper.prepareFinalTables()
        }
    }

    /**
     * Writes [AirbyteRecordMessage] to JDBC database's airbyte_raw table
     *
     * @param database JDBC database to connect to
     * @param sqlOperations interface of SQL queries to execute
     * @param writeConfigs settings for each stream
     * @param catalog catalog of all streams to sync
     */
    private fun recordWriterFunction(
        database: JdbcDatabase,
        sqlOperations: SqlOperations,
        writeConfigs: List<WriteConfig>,
        catalog: ConfiguredAirbyteCatalog,
    ): RecordWriter<PartialAirbyteMessage> {
        var pairToWriteConfig: Map<AirbyteStreamNameNamespacePair, WriteConfig> = emptyMap()

        return RecordWriter {
            pair: AirbyteStreamNameNamespacePair,
            records: List<PartialAirbyteMessage> ->
            if (!pairToWriteConfig.containsKey(pair)) {
                synchronized(JdbcBufferedConsumerFactory) {
                    pairToWriteConfig = writeConfigs.associateBy { toNameNamespacePair(it) }
                }
            }
            require(pairToWriteConfig.containsKey(pair)) {
                String.format(
                    "Message contained record from a stream that was not in the catalog. \ncatalog: %s, \nstream identifier: %s\nkeys: %s",
                    Jsons.serialize(catalog),
                    pair,
                    pairToWriteConfig.keys
                )
            }
            val writeConfig = pairToWriteConfig.getValue(pair)
            sqlOperations.insertRecords(
                database,
                ArrayList(records),
                writeConfig.rawNamespace,
                writeConfig.rawTableName + writeConfig.rawTableSuffix,
                writeConfig.syncId,
                writeConfig.generationId,
            )
        }
    }

    /** Tear down functionality */
    private fun onCloseFunction(
        database: JdbcDatabase,
        sqlOperations: SqlOperations,
        generationIdHandler: JdbcGenerationHandler,
        catalog: ParsedCatalog,
        typerDeduper: TyperDeduper
    ): OnCloseFunction {
        return OnCloseFunction {
            _: Boolean,
            streamSyncSummaries: Map<StreamDescriptor, StreamSyncSummary> ->
            try {
                catalog.streams.forEach {
                    if (
                        it.minimumGenerationId != 0L &&
                            generationIdHandler.getGenerationIdInTable(
                                database,
                                it.id.rawNamespace,
                                it.id.rawName
                            ) != it.generationId &&
                            streamSyncSummaries
                                .getValue(it.id.asStreamDescriptor())
                                .terminalStatus ==
                                AirbyteStreamStatusTraceMessage.AirbyteStreamStatus.COMPLETE
                    ) {
                        sqlOperations.overwriteRawTable(database, it.id.rawNamespace, it.id.rawName)
                    }
                }
                typerDeduper.typeAndDedupe(streamSyncSummaries)
                typerDeduper.commitFinalTables(streamSyncSummaries)
                typerDeduper.cleanup()
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
    }

    private fun toNameNamespacePair(config: WriteConfig): AirbyteStreamNameNamespacePair {
        return AirbyteStreamNameNamespacePair(config.streamName, config.namespace)
    }
}
